src/structs.py

```diff
@@ -303,3 +309,92 @@ class _Option:
     # A termination condition maps a state to a bool, which is True
     # iff the option should terminate now.
     terminal: Callable[[State], bool] = field(repr=False)
+
+
+@dataclass(frozen=True, repr=False, eq=False)
+class Operator:
+    """Struct defining an operator (as in STRIPS). Lifted!
+    """
+    name: str
+    parameters: Sequence[Variable]
+    preconditions: Collection[LiftedAtom]
+    add_effects: Collection[LiftedAtom]
+    delete_effects: Collection[LiftedAtom]
+    option: ParameterizedOption
+    # A sampler maps a state and objects to a option parameters.
+    _sampler: Callable[[State, Sequence[Object]], ArrayLike] = field(repr=False)
+
+    @cached_property
+    def _str(self) -> str:
+        return f"""{self.name}:
+    Parameters: {self.parameters}
+    Preconditions: {self.preconditions}
+    Add Effects: {self.add_effects}
+    Delete Effects: {self.delete_effects}
+    Option: {self.option}"""
+
+    @cached_property
+    def _hash(self) -> int:
+        return hash(str(self))
+
+    def __str__(self) -> str:
+        return self._str
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __eq__(self, other) -> bool:
```

Comment:
Please annotate the type of "other" (actually, I think this is missing in a few places in this file)

Reply:
this is apparently not recommended:

```
predicators/src/structs.py:356: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"
predicators/src/structs.py:356: note: This violates the Liskov substitution principle
predicators/src/structs.py:356: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
predicators/src/structs.py:356: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
predicators/src/structs.py:356: note:     def __eq__(self, other: object) -> bool:
predicators/src/structs.py:356: note:         if not isinstance(other, Operator):
predicators/src/structs.py:356: note:             return NotImplemented
predicators/src/structs.py:356: note:         return <logic to compare two Operator instances>
```